### PR TITLE
Create a separate component out of the new Swipeable row

### PR DESCRIPTION
**Pull request by latekvo**
This PR takes configures `ReanimatedSwipeable` to be a part of the `gesture handler` package.

### CODE REVIEW

**Comment by latekvo** [comment]
https://github.com/software-mansion/react-native-gesture-handler/assets/74246391/7d6a7a5a-47f0-4ea2-9ced-324e7a7ecb6f

**Comment by latekvo** [comment]
This change has to be displayed in the docs, as for most people the reanimated component is likely preferable. Should we do it in this PR (c.c. @kacperkapusciak ), or push this one and open a new one for that?

**Comment by latekvo** [comment]
found a new issue with `forwardRef`'d callbacks i didn't notice before, converting to draft

File: example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx
Comment (j-piasecki) [inline example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx:13]:
```suggestion
} from 'react-native-gesture-handler/ReanimatedSwipeable';
```

File: example/src/new_api/swipeable/GmailStyleSwipeableRow.tsx
Comment (j-piasecki) [inline example/src/new_api/swipeable/GmailStyleSwipeableRow.tsx:13]:
```suggestion
} from 'react-native-gesture-handler/ReanimatedSwipeable';
```

File: example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx
Comment (j-piasecki) [inline example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx:149]: Why set the types explicitly here, shouldn't they be inferred?

And if they are necessary here, why `any`?

File: example/src/new_api/swipeable/GmailStyleSwipeableRow.tsx
Comment (j-piasecki) [inline example/src/new_api/swipeable/GmailStyleSwipeableRow.tsx:90]: Same

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:353]:
```suggestion
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:322]:
```suggestion
         props.onSwipeableWillOpen('left');
```

This function is not a worklet so `runOnJS` isn't needed here.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:324]:
```suggestion
          props.onSwipeableWillOpen('right');
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:327]:
```suggestion
          props.onSwipeableWillClose(closingDirection);
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:336]:
```suggestion
          props.onSwipeableOpen('left', swipeableMethods.current);
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:338]:
```suggestion
          props.onSwipeableOpen('right', swipeableMethods.current);
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:341]:
```suggestion
          props.onSwipeableClose(
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:367]:
```suggestion
              runOnJS(dispatchEndEvents)(fromValue, toValue);
```

Since `dispatchEndEvents` isn't a worklet (and we probably don't want it to be), you need `runOnJS` here.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:372]:
```suggestion
        runOnJS(dispatchImmidiateEvents)(fromValue, toValue);
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:398]:
```suggestion
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:319]:
```suggestion
    const dispatchImmediateEvents = useCallback(
```

File: example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx
Comment (latekvo) [reply example/src/new_api/swipeable/AppleStyleSwipeableRow.tsx:149]: these new eslint rules flag inferring as a bad practice.
`any` is used in cases we don't care about the type of said variable, in our case it's only used when the variable either isn't used, or is printed

File: src/components/ReanimatedSwipeable.tsx
Comment (latekvo) [inline src/components/ReanimatedSwipeable.tsx]: I have to ask about this line.
My linter marks this as:

> unnecessary `prop.onSwipeableWillClose`, `props.onSwipeableWillOpen` deps

But isn't the current setup accommodating changes to those particular fields on the props?
These fields could change without triggering the broad `prop` dep as far as I understand.

Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx]: (in the latest commit, the `props` dep is not missing in the quoted code)

**Comment by latekvo** [comment]
I just found out that I haven't achieved full independence from reanimated after all, I was just using incorrect imports in the examples.

**Comment by latekvo** [comment]
Alright, after some further testing, when exporting `react-native-gesture-handler` the lib works as intended.
`react-native-reanimated` is NOT installed, until user explicitly installs it, even after add `ReanimatedSwipeable` to the project.
But what I don't think is normal nor expected, is `react-native-reanimated` gets installed to the root `node_modules` when running `yarn` within this repo.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [reply src/components/ReanimatedSwipeable.tsx]: I don't get any warnings locally, though I see that the code got changed in the meantime. It could've just lagged behind a bit and a reload would have solved it.

File: tsconfig.json
Comment (j-piasecki) [inline tsconfig.json:28]: I don't think this file exists

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:19]: Similarly to the other PR, I don't think it should be imported from `index`. While in this case, it doesn't introduce any cycles (this component is not imported in `index`), it may turn out to be problematic in the future if someone forgets about it.

**Comment by j-piasecki** [comment]
I'm getting the following error when trying to use the component on an app with RN `0.74.1` and Reanimated `3.12.0`:
Screenshot 2024-06-14 at 09 25 04

I think it may be caused by using `props` directly in methods called on (and from via `runOnJS`) the UI thread. Reanimated will [freeze](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze) everything in the worklets closure, so it may be necessary to only capture what's actually used.

I.e. there's a difference between
```jsx
useAnimatedStyle(() => {
  return { width: props.width }
});
```
and 
```jsx
const width = props.width;
useAnimatedStyle(() => {
  return { width: width }
});
```

File: src/components/ReanimatedSwipeable.tsx
Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx]: After rebuilding i unfortunately still get the following error:
> React Hook useCallback has unnecessary dependencies: 'props.onSwipeableWillClose', 'props.onSwipeableWillOpen', and 'swipeableMethods'. Either exclude them or remove the dependency array.

I think I'll just have to `eslint-disable-next-line` it, because I am quite certain that these are in fact necessary.

**Comment by latekvo** [comment]
Besides the built-in examples issue, when using this library as usual, everything is now tested and working. Couldn't find any discrepancies so far when testing side by side with the original, but I still have to check the exact values returned by the callback and rendering functions, as they are now computed by a completely different mechanism.

**Comment by latekvo** [comment]
i believe all the issues i could find are now fixed, besides a slight difference in spring config, but as far as i looked, there isn't a simple way to convert the legacy spring config 1:1 to a new one

**Comment by j-piasecki** [comment]
There's an issue with layout in the new example (on all platforms):

Screenshot 2024-06-18 at 08 40 46

**Comment by latekvo** [comment]
### How did you achieve that? I replicate your issue.

https://github.com/software-mansion/react-native-gesture-handler/assets/74246391/f37ab86d-d0f6-4e37-b4b2-112bd83b68b8


Uploading Screen Recording 2024-06-18 at 09.15.58.mov…

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (j-piasecki) [inline example/src/release_tests/swipeableReanimation/index.tsx:2]: Is there a story behind `Laminated` 😄?

Anyway, If you need to use both Animated and Reanimated in the same file, I'd stick to keeping Animated as Animated, and import Reanimated as Reanimated.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:641]: I don't think this is needed here (and it was causing the issue from the screen above)

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:606]: This way you cannot 'swipe back' the swipeable. I think that pan should be attached to the outer `AnimatedView`, and tap to the inner one.

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (j-piasecki) [inline example/src/release_tests/swipeableReanimation/index.tsx:10]:
```suggestion
import ReanimatedSwipeable from 'react-native-gesture-handler/ReanimatedSwipeable';
```

Doesn't this work?

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (latekvo) [reply example/src/release_tests/swipeableReanimation/index.tsx:2]: I'm sorry, i never thought i'd upload that example and forgot to change the names 🫢

File: src/components/ReanimatedSwipeable.tsx
Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx:641]: It ensures the element is clickable and draggable across it's entire area, 
but I think you're right suggest to removing it, as the wasn't clickable across the entire area.

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (latekvo) [reply example/src/release_tests/swipeableReanimation/index.tsx:10]: In the file, it's exported as `Swipeable`, to be a drop-in replacement.
I can change it's name to `ReanimatedSwipeable` if you think that would work better.

Comment (j-piasecki) [reply example/src/release_tests/swipeableReanimation/index.tsx:10]: It's a default export, you don't have to import it under the same name it was exported (and you don't need to use `as` to rename).

What you did would be necessary if it was a named export:
```jsx
export { ReanimatedSwipeable }
```
vs
```jsx
export default ReanimatedSwipeable
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:602]:
```suggestion
          <GestureDetector gesture={tapGesture}>
```

That's not necessary, it basically wouldn't do anything since gestures are configured in the builder in the new API.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:593]:
```suggestion
```

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (latekvo) [reply example/src/release_tests/swipeableReanimation/index.tsx]: I see, didn't understand at first. I had no idea that's how default exports work.

**Comment by latekvo** [comment]
left swipe threshold is twice as large as it should be on both the new and the legacy version, will double check, but looks like it

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (j-piasecki) [inline example/src/release_tests/swipeableReanimation/index.tsx:15]:
```suggestion
import Reanimated, {
  SharedValue,
  useAnimatedStyle,
} from 'react-native-reanimated';
```

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:583]:
```suggestion
```

That's a bad idea - you're effectively removing the config set above to assign props to it. The only prop that it would make sense for is `enableTrackpadTwoFingerGesture` and you can just set it on the gesture directly:
```jsx
panGesture.enableTrackpadTwoFingerGesture(props.enableTrackpadTwoFingerGesture)
```

Also you can remove `useNativeAnimations` from `SwipeableProps` since all Reanimated animations are native.

Comment (latekvo) [reply src/components/ReanimatedSwipeable.tsx]: I think it wasn't so bad, it automatically recognized all suitable props, making room for change in the future as well, and applied them before any of them were explicitly set. It's also quite readable.
But I replaced it as per your recommendation.

**Comment by latekvo** [comment]
Just found some weird behaviour on `IOS` devices :/
`Taps` aren't detect and `swipe-back` doesn't register.

File: src/components/ReanimatedSwipeable.tsx
Comment (j-piasecki) [inline src/components/ReanimatedSwipeable.tsx:605]:
```suggestion
          <GestureDetector gesture={tapGesture} touchAction="pan-y">
```

**Review by j-piasecki — APPROVED** [review]
Looks good 🚀.

I left one comment (fixes scrolling using touch on web), please wait for @m-bert's review before merging.

File: src/components/ReanimatedSwipeable.tsx
Comment (m-bert) [inline src/components/ReanimatedSwipeable.tsx:331]: According to our conversation, this should be removed

```suggestion
```

File: src/components/ReanimatedSwipeable.tsx
Comment (m-bert) [inline src/components/ReanimatedSwipeable.tsx:351]: Same as above
```suggestion
```

File: example/src/release_tests/swipeableReanimation/index.tsx
Comment (m-bert) [inline example/src/release_tests/swipeableReanimation/index.tsx:61]: It would be nice to have some feedback if we use `RectButton` (could be for example `console.log`). Also, why legacy version uses `button` if the one based on `Reanimated` doesn't?

**Review by m-bert — CHANGES_REQUESTED** [review]
Imho this PR has too many linter changes that are out of scope (like `tsconfig.json` inside `MacOSExample`). It would be better to do another PR that unifies all indentations etc. (cc @j-piasecki). I don't consider it as a blocker though.

What I do consider as a blocker is lack of `touchAction: "pan-y"` in `GestureDetector` with `tapGesture`. Without it, `View` will have `touch-action: none;` which will effectively disable scrolling (you can check that on code from [this PR](https://github.com/software-mansion/react-native-gesture-handler/pull/2788)).

Also, please remove `eslint-disable` (as we discussed it shouldn't be necessary here).

Other than that it looks solid, good job! 😄

File: example/src/new_api/swipeable/index.tsx
Comment (latekvo) [reply example/src/new_api/swipeable/index.tsx:51]: Linter marks inline components like this as a bad practice.

**Review by m-bert — APPROVED** [review]
Thanks for adding `touchAction` 😄

I still believe that it would be nice to have some feedback from `RectButtons` in new example (using them as `View` doesn't make much sense for me), so please either remove them or add a callback before merging 😅

**Comment by latekvo** [comment]
> I still believe that it would be nice to have some feedback from `RectButtons` in new example (using them as `View` doesn't make much sense for me), so please either remove them or add a callback before merging 😅

I removed those completely, they were used in the apple/gmail swipeable showcase, and I must have copied them accidentally.

**Comment by jackstudd** [comment]
Hi @latekvo !🙋‍♂️

Thanks for this! It works really great and with better performances! However, I noticed there's no longer the ability to disable the swipe. I'm guessing this could be done with a props and passed to the `Gesture.Pan.enabled()`.

I tried to just remove the `renderActions` based on my disabled condition, but we can still swipe, we just no longer see the action.

I would gladly do the PR myself if needed!

**Comment by jackstudd** [comment]
Hey @latekvo !

Just tried it out, it works perfectly, many thanks!

**Pull request merged**
